### ISSUE TITLE
Sentences (do not) need to be shuffled

### ISSUE THREAD

**Issue by alvelvis** [UniversalDependencies/UD_Portuguese-PetroGold#3]
Hi, this is for the release team (@dan-zeman, who else?)

Sentences in PetroGold-UD files should be shuffled, since it has only 19 (big) academic documents, thus same documents will have one same author, writing style and subject. By mistake, sentences in `pt_petrogold-ud-train.conllu` are not shuffled, while the other files have shuffled sentences.

Can I reorder the sentences in each file for the next release? I know it should be avoided, so it will hopefully not happen again.

**Comment by martinpopel** [comment]
[UD recommends](https://universaldependencies.org/release_checklist.html#data-split):

> Shuffling sentences should be avoided if possible, but sometimes it is necessary in order to prevent copyright issues. If you must shuffle, consider shuffling blocks of sentences (up to N characters long) rather than individual sentences.

If there is no copyright issue, don't shuffle any sentences in train/dev/test. This allows to add cross-sentence annotations (e.g. coreference) in future or parsers/taggers may get better results when considering the following/previous sentences etc.

> thus same documents will have one same author, writing style and subject.

I don't understand what is the problem here.

**Comment by alvelvis** [comment]
Hi @martinpopel

These are good reasons not to shuffle. My point is that by not shuffling sentences we are training models based on few documents, so there is not enough diversity in the training data (much less diversity in test/dev data). Is it irrelevant compared to the benefits of not shuffling sentences?

**Comment by arademaker** [comment]
@alvelvis I have recently moved to group sentences into documents in the dev/test/train. You can see the discussion and the @manning arguments [here](https://github.com/UniversalDependencies/UD_Portuguese-Bosque/issues/203#issuecomment-956369840).

**Comment by martinpopel** [comment]
One of the strong UD recommendations is to keep the data split of one treebank stable across UD releases, i.e. make sure a sentence that was in training data in release N is not moved to dev or test data in release N+1. If you now split one document between train and test, then there would be almost no way to put it together in future (you could put it whole into train, but then you would need another document for the new test set).

If the whole treebank has just 19 authors (documents), you can put e.g. 15 into train, 2 into test and 2 into dev (this seems to be closest to the recommended 80/10/10 split, if the documents are approximately of the same length). I agree the author/style diversity is not great, but having 19 instead of 15 does not seem as such a big difference. Also, if there is no overlap in authors between the train and test sets, the results will be more realistic (it seems unlikely that the parser's goal is to be applied only to texts by the 19 authors only).

**Comment by martinpopel** [comment]
BTW: if there are 19 documents, it would be useful to annotate their boundaries (and possibly also paragraph boundaries) in [the UD recommended way](https://universaldependencies.org/format.html#paragraph-and-document-boundaries), i.e. adding lines such as 
`# newdoc id = 47-20140910-MONOGRAFIA`
before the first line of each document.

**Comment by amir-zeldes** [comment]
> Also, if there is no overlap in authors between the train and test sets, the results will be more realistic

I just wanted to add that we have seen this repeatedly in experiments: test numbers on datasets with shuffled sentences crossing train/test partitions are consitently over-optimistic compared to truly unseen data, due to data-leakage between the partitions (vocabulary and other syntactic choices are improbably similar across train-test).

**Comment by alvelvis** [comment]
Thanks for the many advices, everyone.

> @alvelvis I have recently moved to group sentences into documents in the dev/test/train. You can see the discussion and the @manning arguments [here](https://github.com/UniversalDependencies/UD_Portuguese-Bosque/issues/203#issuecomment-956369840).

@arademaker I like the way you re-split the sentences in Bosque, and the arguments from @manning in the cited mail list are great, thanks for sharing them.

> If the whole treebank has just 19 authors (documents), you can put e.g. 15 into train, 2 into test and 2 into dev (this seems to be closest to the recommended 80/10/10 split, if the documents are approximately of the same length). I agree the author/style diversity is not great, but having 19 instead of 15 does not seem as such a big difference. Also, if there is no overlap in authors between the train and test sets, the results will be more realistic (it seems unlikely that the parser's goal is to be applied only to texts by the 19 authors only).

@martinpopel I think I'll do something similar to your suggestion: I tested and the 19 documents do not have similar length, but I managed to organize them in a way that train will have 80,1% of sentences (15 entire documents), test has 11,6% (2 entire documents) and dev 8,2% (2 entire documents). There are no overlapping documents among the files, so no overlapping sentences as well. And the sentences are not shuffled. Dev and test will have slight different sizes but I think it's for a greater good: keeping entire documents in each file.

> BTW: if there are 19 documents, it would be useful to annotate their boundaries (and possibly also paragraph boundaries) in [the UD recommended way](https://universaldependencies.org/format.html#paragraph-and-document-boundaries), i.e. adding lines such as `# newdoc id = 47-20140910-MONOGRAFIA` before the first line of each document.

I'm adding this now, thanks! Paragraph is an information we do not have, though.

> I just wanted to add that we have seen this repeatedly in experiments: test numbers on datasets with shuffled sentences crossing train/test partitions are consitently over-optimistic compared to truly unseen data, due to data-leakage between the partitions (vocabulary and other syntactic choices are improbably similar across train-test).

@amir-zeldes Really interesting, it sounded counterintuitive to me at first but reading the whole discussion now makes sense.
